### DISCUSSION TITLE
Callback being executed twice when using click.option and prompt

### DISCUSSION THREAD

**Discussion by MaciekKrz** [pallets/click#2673] [Q&A] [answered]
When I have a situation like below:

`@click.option(
    "--redirect_uris", required=True, prompt=True, callback=url_list_type, help="list of redirect uris")`

My callback function is being executed twice and that causes errors. 
Is there any solution I can implement to avoid it?

**Comment by davidism** [comment, marked as answer]
The callback is used both to validate the prompt input, and then independently from that is used to validate the value passed to the option (which in this case came from the prompt). Usually this shouldn't matter, but if you're doing something expensive it might. You could omit the prompt from the option, and instead call `prompt` yourself in the command if the option wasn't given.

**Reply by MaciekKrz** [reply]
Thanks for the tip, David.
I see now that the callback runs twice: once for the prompt and once for the option. I'm thinking of either removing the prompt from the option and adding a manual prompt in the command as you suggested or finding a way to check in the callback if it's already been processed. 

Either way, I've got a clearer idea of what to do now. Thanks again!

**Comment by MaciekKrz** [comment]
Hi,
I didn't solve it in the click itself.
I've just added some checks in the callback/validator to make it work...

[image: Zrzut ekranu 2024-08-27 o 10.40.06.png]

pon., 26 sie 2024 o 14:41 yorickdowne ***@***.***> napisał(a):

**Reply by yorickdowne** [reply]
Thanks! I was not able to figure out how to see whether the callback was being called for prompt validation or value validation, and so moved the prompt into the callback.

PR for that at https://github.com/eth-educators/ethstaker-deposit-cli/pull/98 if someone else has a similar issue and wants sample code.
